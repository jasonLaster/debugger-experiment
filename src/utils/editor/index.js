const { isPretty, isJavaScript } = require("../source");
const { isOriginalId } = require("../source-map");
const buildQuery = require("./build-query");
const {
  getDocument,
  setDocument,
  removeDocument,
  clearDocuments
} = require("./source-documents");

const {
  countMatches,
  find,
  findNext,
  findPrev,
  removeOverlay,
  clearIndex
} = require("./source-search");

const SourceEditor = require("./source-editor");

function shouldShowPrettyPrint(selectedSource) {
  if (!selectedSource) {
    return false;
  }

  selectedSource = selectedSource.toJS();
  const _isPretty = isPretty(selectedSource);
  const _isJavaScript = isJavaScript(selectedSource.url);
  const isOriginal = isOriginalId(selectedSource.id);
  const hasSourceMap = selectedSource.sourceMapURL;

  if (_isPretty || isOriginal || hasSourceMap || !_isJavaScript) {
    return false;
  }

  return true;
}

function onKeyDown(codeMirror, e) {
  let { key, target } = e;
  let codeWrapper = codeMirror.getWrapperElement();
  let textArea = codeWrapper.querySelector("textArea");

  if (key === "Escape" && target == textArea) {
    e.stopPropagation();
    e.preventDefault();
    codeWrapper.focus();
  } else if (key === "Enter" && target == codeWrapper) {
    e.preventDefault();
    // Focus into editor's text area
    textArea.focus();
  }
}

function shouldShowFooter(selectedSource, horizontal) {
  if (!horizontal) {
    return true;
  }

  return shouldShowPrettyPrint(selectedSource);
}

function forEachLine(codeMirror, iter) {
  codeMirror.doc.iter(0, codeMirror.lineCount(), iter);
}

function removeLineClass(codeMirror, line, className) {
  codeMirror.removeLineClass(line, "line", className);
}

function clearLineClass(codeMirror, className) {
  forEachLine(codeMirror, line => {
    removeLineClass(codeMirror, line, className);
  });
}

function isTextForSource(sourceText) {
  return !sourceText.get("loading") && !sourceText.get("error");
}

function breakpointAtLine(breakpoints, line) {
  return breakpoints.find(b => {
    return b.location.line === line + 1;
  });
}

function getTextForLine(codeMirror, line) {
  return codeMirror.getLine(line - 1).trim();
}

function getCursorLine(codeMirror) {
  return codeMirror.getCursor().line;
}
/**
 * Forces the breakpoint gutter to be the same size as the line
 * numbers gutter. Editor CSS will absolutely position the gutter
 * beneath the line numbers. This makes it easy to be flexible with
 * how we overlay breakpoints.
 */
function resizeBreakpointGutter(editor) {
  const gutters = editor.display.gutters;
  const lineNumbers = gutters.querySelector(".CodeMirror-linenumbers");
  const breakpoints = gutters.querySelector(".breakpoints");
  breakpoints.style.width = `${lineNumbers.clientWidth}px`;
}

function traverseResults(e, ctx, query, dir, modifiers) {
  e.stopPropagation();
  e.preventDefault();

  if (dir == "prev") {
    findPrev(ctx, query, true, modifiers);
  } else if (dir == "next") {
    findNext(ctx, query, true, modifiers);
  }
}

function createEditor() {
  return new SourceEditor({
    mode: "javascript",
    readOnly: true,
    lineNumbers: true,
    theme: "mozilla",
    lineWrapping: false,
    matchBrackets: true,
    showAnnotationRuler: true,
    enableCodeFolding: false,
    gutters: ["breakpoints", "hit-markers"],
    value: " ",
    extraKeys: {
      // Override code mirror keymap to avoid conflicts with split console.
      Esc: false,
      "Cmd-F": false,
      "Cmd-G": false
    }
  });
}

/**
 * Disable APZ for source editors. It currently causes the line numbers to
 * "tear off" and swim around on top of the content. Bug 1160601 tracks
 * finding a solution that allows APZ to work with CodeMirror.
 */
function onWheel(cm, ev) {
  // By handling the wheel events ourselves, we force the platform to
  // scroll synchronously, like it did before APZ. However, we lose smooth
  // scrolling for users with mouse wheels. This seems acceptible vs.
  // doing nothing and letting the gutter slide around.
  ev.preventDefault();

  let { deltaX, deltaY } = ev;

  if (ev.deltaMode == ev.DOM_DELTA_LINE) {
    deltaX *= cm.defaultCharWidth();
    deltaY *= cm.defaultTextHeight();
  } else if (ev.deltaMode == ev.DOM_DELTA_PAGE) {
    deltaX *= cm.getWrapperElement().clientWidth;
    deltaY *= cm.getWrapperElement().clientHeight;
  }

  cm.getScrollerElement().scrollBy(deltaX, deltaY);
}

module.exports = {
  createEditor,
  onWheel,
  shouldShowPrettyPrint,
  shouldShowFooter,
  clearLineClass,
  onKeyDown,
  buildQuery,
  getDocument,
  setDocument,
  removeDocument,
  clearDocuments,
  countMatches,
  find,
  findNext,
  findPrev,
  clearIndex,
  removeOverlay,
  isTextForSource,
  breakpointAtLine,
  getTextForLine,
  getCursorLine,
  resizeBreakpointGutter,
  traverseResults
};
