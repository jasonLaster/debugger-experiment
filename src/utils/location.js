import type { Location, SourceId } from "debugger-html";

type IncompleteLocation = {
  sourceId: SourceId,
  line: ?number,
  column: ?number,
  sourceUrl?: string
};

export function createLocation({
  sourceId,
  line,
  column,
  sourceUrl
}: IncompleteLocation): Location {
  return {
    sourceId,
    line: line,
    column: column,
    sourceUrl: sourceUrl || null
  };
}
