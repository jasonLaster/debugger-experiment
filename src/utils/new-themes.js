import themes from "devtools-themes";

window.updateTheme = function(theme) {
  const root = document.body.parentNode;
  const appRoot = document.querySelector(".launchpad-root");

  root.className = "";
  appRoot.className = "launchpad-root";

  root.classList.add(`theme-${theme}`);
  appRoot.classList.add(`theme-${theme}`);
};

window.themeNames = Object.keys(themes);
window.updateThemes = function() {
  window.themeNames.forEach((theme, i) =>
    setTimeout(() => window.updateTheme(theme), i * 1000)
  );
};

window.setTimeout(() => {
  // remove the stylesheet where the original variables are defined
  // because we'll be assigning the variables to base16 definitions
  // e.g. :root .dark-theme { --theme-comment: #fff }
  document.styleSheets[2].ownerNode.remove();
  window.updateTheme("dracula");
}, 1000);
