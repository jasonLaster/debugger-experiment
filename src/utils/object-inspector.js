
if (typeof window == "object") {

}
const WINDOW_PROPERTIES = Object.getOwnPropertyNames(window || {});

function nodeHasChildren(item) {
  return Array.isArray(item.contents);
}

function nodeIsOptimizedOut(item) {
  return !nodeHasChildren(item) && item.contents.value.optimizedOut === true;
}

function nodeIsMissingArguments(item) {
  return !nodeHasChildren(item) &&
    item.contents.value.missingArguments === true;
}

function nodeHasProperties(item) {
  return !nodeHasChildren(item) && item.contents.value.type === "object";
}

function nodeIsPrimitive(item) {
  return !nodeHasChildren(item) && !nodeHasProperties(item);
}

function isDefault(item) {
  return WINDOW_PROPERTIES.includes(item.name);
}

function makeNodesForProperties(objProps, parentPath) {
  const { ownProperties, prototype } = objProps;

  const nodes = Object.keys(ownProperties).sort().filter(name => {
    // Ignore non-concrete values like getters and setters
    // for now by making sure we have a value.
    return "value" in ownProperties[name];
  });

  if (nodes.length > 100) {
    makeBuckets;
    nodes.map(name => {
      return createNode(name, `${parentPath}/${bucket}/${name}`, ownProperties[name]);
    });
  } else {
    nodes.map(name => {
      return createNode(name, `${parentPath}/${name}`, ownProperties[name]);
    });
  }

  // Add the prototype if it exists and is not null
  if (prototype && prototype.type !== "null") {
    nodes.push(createNode(
      "__proto__",
      `${parentPath}/__proto__`,
      { value: prototype }
    ));
  }

  return nodes;
}

function createNode(name, path, contents) {
  // The path is important to uniquely identify the item in the entire
  // tree. This helps debugging & optimizes React's rendering of large
  // lists. The path will be separated by property name,
  // i.e. `{ foo: { bar: { baz: 5 }}}` will have a path of `foo/bar/baz`
  // for the inner object.
  return { name, path, contents };
}

module.exports = {
  nodeHasChildren,
  makeNodesForProperties
};
