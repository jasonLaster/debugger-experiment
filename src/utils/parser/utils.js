// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const isEmpty = require("lodash/isEmpty");
const uniq = require("lodash/uniq");
import getAst, { traverseAst } from "./getAst";
import { isFunction, isVariable } from "./shared";
import type { SourceText, Location, Frame, TokenResolution } from "../../types";

function getNodeValue(node) {
  if (t.isThisExpression(node)) {
    return "this";
  }

  return node.name;
}

function getScopeVariables(scope: Scope) {
  const { bindings } = scope;

  return toPairs(bindings).map(([name, binding]) => ({
    name,
    references: binding.referencePaths
  }));
}

function getScopeChain(scope: Scope): Scope[] {
  const scopes = [scope];

  do {
    scopes.push(scope);
  } while ((scope = scope.parent));

  return scopes;
}

function isLexicalScope(path) {
  return isFunction(path) || t.isProgram(path);
}

// Resolves a token (at location) in the source to determine if it is in scope
// of the given frame and the expression (if any) to which it belongs
export function resolveToken(
  source: SourceText,
  token: string,
  location: Location,
  frame: Frame
): ?TokenResolution {
  const expression = getClosestExpression(source, token, location);
  const scope = getClosestScope(source, location);

  if (!expression || !expression.value || !scope) {
    return { expression: null, inScope: false };
  }

  const inScope = isExpressionInScope(expression.value, scope);

  return {
    expression,
    inScope
  };
}

export function getVariablesInLocalScope(scope: Scope) {
  return getScopeVariables(scope);
}

export function getVariablesInScope(scope: Scope) {
  const scopes = getScopeChain(scope);
  const scopeVars = scopes.map(getScopeVariables);
  const vars = [{ name: "this" }, { name: "arguments" }]
    .concat(...scopeVars)
    .map(variable => variable.name);
  return uniq(vars);
}

export function isExpressionInScope(expression: string, scope?: Scope) {
  if (!scope) {
    return false;
  }

  const variables = getVariablesInScope(scope);
  const firstPart = expression.split(/\./)[0];
  return variables.includes(firstPart);
}
