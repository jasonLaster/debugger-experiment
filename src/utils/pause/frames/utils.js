/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

export function formatCopyName(frame: LocalFrame) {
  const displayName = formatDisplayName(frame);
  const fileName = getFilename(frame.source);
  const frameLocation = frame.location.line;

  return `${displayName} (${fileName}#${frameLocation})`;
}
