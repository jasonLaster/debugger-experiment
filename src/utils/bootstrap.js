/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import React from "react";
import { bindActionCreators, combineReducers } from "redux";
import ReactDOM from "react-dom";
const { Provider } = require("react-redux");

import { isFirefoxPanel, isDevelopment, isTesting } from "devtools-environment";
import { startSourceMapWorker, stopSourceMapWorker } from "devtools-source-map";
import * as search from "../workers/search";
import * as prettyPrint from "../workers/pretty-print";
import * as parser from "../workers/parser";

import configureStore from "../actions/utils/create-store";
import reducers from "../reducers";
import * as selectors from "../selectors";
import App from "../components/App";
import { prefs } from "./prefs";

function renderPanel(component, store) {
  const root = document.createElement("div");
  root.className = "launchpad-root theme-body";
  root.style.setProperty("flex", "1");
  const mount = document.querySelector("#mount");
  if (!mount) {
    return;
  }
  mount.appendChild(root);

  ReactDOM.render(
    React.createElement(Provider, { store }, React.createElement(component)),
    root
  );
}

export function bootstrapStore(client: any, { services, toolboxActions }: any) {
  const createStore = configureStore({
    log: isTesting(),
    timing: isDevelopment(),
    makeThunkArgs: (args, state) => {
      return { ...args, client, ...services, ...toolboxActions };
    }
  });

  const store = createStore(combineReducers(reducers));
  store.subscribe(() => updatePrefs(store.getState()));

  const actions = bindActionCreators(
    require("../actions").default,
    store.dispatch
  );

  return { store, actions, selectors };
}

export function bootstrapWorkers() {
  const workerPath = isDevelopment()
    ? "assets/build"
    : "resource://devtools/client/debugger/new/dist";
  const sourceMaps = require("devtools-source-map");
  const prettyPrint = require("../workers/pretty-print");
  const parser = require("../workers/parser");
  const search = require("../workers/search");

  return { sourceMaps, prettyPrint, parser, search };
}

export function teardownWorkers() {}

export function bootstrapApp(store: any) {
  if (isFirefoxPanel()) {
    renderPanel(App, store);
  } else {
    const { renderRoot } = require("devtools-launchpad");
    renderRoot(React, ReactDOM, App, store);
  }
}

function updatePrefs(state: any) {
  const pendingBreakpoints = selectors.getPendingBreakpoints(state);

  if (prefs.pendingBreakpoints !== pendingBreakpoints) {
    prefs.pendingBreakpoints = pendingBreakpoints;
  }
}
