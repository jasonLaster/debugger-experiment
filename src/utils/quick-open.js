/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import { endTruncateStr } from "./utils";
import { isPretty, getSourcePath } from "./source";

import type { Location as BabelLocation } from "@babel/types";
import type { SourcesMap } from "../reducers/sources";
import type { Symbols } from "../reducers/ast";
import type { QuickOpenType } from "../reducers/quick-open";
import type { SymbolDeclaration } from "../workers/parser";
import type { Source } from "../types";

export const MODIFIERS = {
  "@": "functions",
  "#": "variables",
  ":": "goto",
  "?": "shortcuts"
};

export function parseQuickOpenQuery(query: string): QuickOpenType {
  const modifierPattern = /^@|#|:|\?$/;
  const gotoSourcePattern = /^(\w+)\:/;
  const startsWithModifier = modifierPattern.test(query[0]);
  const isGotoSource = gotoSourcePattern.test(query);

  if (startsWithModifier) {
    const modifier = query[0];
    return MODIFIERS[modifier];
  }

  if (isGotoSource) {
    return "gotoSource";
  }

  return "sources";
}

export function parseLineColumn(query: string) {
  const [, line, column] = query.split(":");
  const lineNumber = parseInt(line, 10);
  const columnNumber = parseInt(column, 10);
  if (!isNaN(lineNumber)) {
    return {
      line: lineNumber,
      ...(!isNaN(columnNumber) ? { column: columnNumber } : null)
    };
  }
}

export function formatSourcesForProjectDirectoryRoot(
  source: Source,
  projectDirectoryRoot: string
) {
  const sourcePath = getSourcePath(source.get("url"));
  let title = "";
  let subtitle = "";

  if (sourcePath) {
    const sourcePathSplit = sourcePath.split("/");

    // Remove leading "/"
    if (sourcePathSplit[0] === "") {
      sourcePathSplit.shift();
    }

    // Make the title the file name
    title = sourcePathSplit.pop().split("?")[0];

    // Rebuild the sourcePath
    let newSourcePath = sourcePathSplit.join("/");

    if (projectDirectoryRoot != "") {
      // Remove the domain from the source path
      const projectDirectoryRootSplit = projectDirectoryRoot.split("/");
      // Remove the first item, which is "/"
      projectDirectoryRootSplit.shift();
      // Remove the domain
      projectDirectoryRootSplit.shift();

      // Remove the directory root, if it matches, from the source path
      const projectDirectoryJoined = projectDirectoryRootSplit.join("/");
      if (newSourcePath.includes(projectDirectoryJoined)) {
        newSourcePath = newSourcePath.replace(projectDirectoryJoined, "");
      }
    }

    // Make the subtitle the remaining info
    subtitle = endTruncateStr(newSourcePath, 100);
  }

  return {
    value: sourcePath,
    title,
    subtitle,
    id: source.get("id"),
    url: source.get("url")
  };
}

export type QuickOpenResult = {|
  id: string,
  value: string,
  title: string,
  subtitle?: string,
  location?: BabelLocation,
  url?: string
|};

export type FormattedSymbolDeclarations = {|
  variables: Array<QuickOpenResult>,
  functions: Array<QuickOpenResult>
|};

export function formatSymbol(symbol: SymbolDeclaration): QuickOpenResult {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

export function formatSymbols(symbols: ?Symbols): FormattedSymbolDeclarations {
  if (!symbols || symbols.loading) {
    return { variables: [], functions: [] };
  }

  const { variables, functions } = symbols;

  return {
    variables: variables.map(formatSymbol),
    functions: functions.map(formatSymbol)
  };
}

export function formatShortcutResults(): Array<QuickOpenResult> {
  return [
    {
      value: L10N.getStr("symbolSearch.search.functionsPlaceholder.title"),
      title: `@ ${L10N.getStr("symbolSearch.search.functionsPlaceholder")}`,
      id: "@"
    },
    {
      value: L10N.getStr("symbolSearch.search.variablesPlaceholder.title"),
      title: `# ${L10N.getStr("symbolSearch.search.variablesPlaceholder")}`,
      id: "#"
    },
    {
      value: L10N.getStr("gotoLineModal.title"),
      title: `: ${L10N.getStr("gotoLineModal.placeholder")}`,
      id: ":"
    }
  ];
}

export function formatSources(
  sources: SourcesMap,
  root: string
): Array<QuickOpenResult> {
  return sources
    .valueSeq()
    .filter(source => !isPretty(source))
    .map(source => formatSourcesForProjectDirectoryRoot(source, root))
    .filter(({ value }) => value != "")
    .toJS();
}
