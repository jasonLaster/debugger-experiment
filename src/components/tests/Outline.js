import { createFactory } from "react";
import { shallow } from "enzyme";

import _Outline from "../../components/Outline";
const Outline = createFactory(_Outline.WrappedComponent);
import { setValue } from "devtools-config";

const fromJS = require("../../utils/fromJS");
import { getSourceText } from "../../test/unit/utils";

describe("Outline", () => {
  it("should render", () => {
    setValue("features.outline", true);
    const sourceText = fromJS(getSourceText("func"));
    const wrapper = shallow(
      new Outline({
        sourceText
      })
    );

    expect(wrapper).toMatchSnapshot();
  });
});
