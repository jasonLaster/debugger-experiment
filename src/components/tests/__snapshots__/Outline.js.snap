// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outline should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="outline"
>
    <ul />
</div>,
  "nodes": Array [
    <div
      className="outline"
>
      <ul />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Outline
        sourceText={
                Object {
                        "contentType": "text/javascript",
                        "id": "func",
                        "text": "function square(n) {
                  return n * n;
                }
                
                child = function() {};
                
                (function() {
                  2;
                })();
                ",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Outline {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "sourceText": Object {
            "contentType": "text/javascript",
            "id": "func",
            "text": "function square(n) {
  return n * n;
}

child = function() {};

(function() {
  2;
})();
",
          },
        },
        "refs": Object {},
        "state": Object {},
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="outline"
>
          <ul />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="outline"
>
          <ul />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Outline
    sourceText={
        Object {
            "contentType": "text/javascript",
            "id": "func",
            "text": "function square(n) {
          return n * n;
        }
        
        child = function() {};
        
        (function() {
          2;
        })();
        ",
          }
    }
/>,
}
`;
